src/Components/Endpoints/gen/ComponentEndpointsGenerator.cs

```diff
@@ -0,0 +1,478 @@
+
+        var getComponentsSignature = assembliesReferencingComponents.Select(
+            static (assembly, ct) => (assembly, signature: Emitter.CreateGetComponentsMethodSignature(assembly)));
+
+        var getComponentsBodyThunk = referencedAssembliesComponents
+            .Select(static (cm, ct) => (cm.Component.ContainingAssembly, body: Emitter.GetComponentsBody(cm)));
+
+        var groupedComponentStatements = getComponentsBodyThunk
+            .Collect()
+            .Select(static (gpbt, ct) => gpbt.GroupBy(kvp => kvp.ContainingAssembly, SymbolEqualityComparer.Default)
+                .ToDictionary(
+                    kvp => kvp.Key,
+                    kvp => kvp.Select(t => t.body).ToImmutableArray(), SymbolEqualityComparer.Default));
+
+        var getComponentMethodThunks = getComponentsSignature
+            .Combine(groupedComponentStatements)
+            .Select(static (ctx, ct) =>
+            {
+                var (assembly, signature) = ctx.Left;
+                var bodyStatements = ctx.Right;
+                var body = bodyStatements[assembly];
+                return Emitter.CreateGetMethod(signature, body);
+            });
+
+        return (projectGetComponentComponentsBodyThunk, getComponentMethodThunks);
+    }
+
+    private static IEnumerable<ComponentModel> ComponentWithAssembly(
+        (((IAssemblySymbol, INamedTypeSymbol), INamedTypeSymbol), INamedTypeSymbol) context,
+        CancellationToken _)
+    {
+        var (((candidate, componentsInterface), routeAttribute), renderModeAttribute) = context;
+        var module = candidate.Modules.Single();
+
+        var componentCollector = new ComponentCollector(componentsInterface, routeAttribute, renderModeAttribute);
+        componentCollector.Visit(module.GlobalNamespace);
+
+        foreach (var item in componentCollector.Components!)
+        {
+            yield return item;
+        }
+    }
+
+    private ComponentModel CreateComponentModel((((ISymbol?, INamedTypeSymbol), INamedTypeSymbol), INamedTypeSymbol) context, CancellationToken token)
+    {
+        var (((component, _), routeAttribute), renderModeAttribute) = context;
+        return ComponentModel.FromType((INamedTypeSymbol)component!, routeAttribute, renderModeAttribute);
+    }
+
+    private bool IsComponent((ISymbol? candidate, INamedTypeSymbol componentInterface) tuple)
+    {
+        return tuple.candidate is INamedTypeSymbol componentType &&
+            ComponentCollector.IsComponent(componentType, tuple.componentInterface);
+    }
+
+    private bool FilterAssemblies((IAssemblySymbol assembly, IAssemblySymbol componentsAssembly) context)
+    {
+        var (assembly, componentsAssembly) = context;
+        if (assembly.Name.StartsWith("System.", StringComparison.Ordinal) ||
+            assembly.Name.StartsWith("Microsoft.", StringComparison.Ordinal))
+        {
+            // Filter out system assemblies as well as our components assemblies.
+            return false;
```

Comment:
Is this only being used for routable page components? Just want to understand why we would be excluding our own framework components. If this is for build perf, another approach to consider would be filtering out assemblies that don't transitively reference M.A.Components, as that would eliminate everything from System.*.

Reply:
It was done for perf, but on second thought this looks like premature optimization on my part, so I am going to remove it.

Assembly references are not transitive. They are always a flat list, hence why I wanted to avoid the excessive scanning, but in retrospective, this will only happen once, so I think it's ok.